File: app/src/main/java/com/monk/rxjava2/imetate2/DisposableHelper.java
package com.monk.rxjava2.imetate2;

import java.util.concurrent.atomic.AtomicReference;

/**
 * @author monk
 * @date 2019-03-01
 */
public enum DisposableHelper implements Disposable {
    /*** */
    DISPOSABLE;

    public static boolean isDisposed(Disposable disposable) {
        return disposable == DISPOSABLE;
    }

    public static boolean dispose(AtomicReference<Disposable> disposableAtomicReference) {
        Disposable current = disposableAtomicReference.get();
        Disposable d = DISPOSABLE;
        if (current != d) {
            current = disposableAtomicReference.getAndSet(d);
            if (current != d) {
                if (current != null) {
                    current.dispose();
                }
                return true;
            }
        }
        return false;
    }


    @Override
    public boolean isDisposed() {
        return false;
    }

    @Override
    public void dispose() {

    }
}

File: module_fragment/src/main/java/com/monk/modulefragment/rxjava2/imetate2/DisposableHelper.kt
package com.monk.modulefragment.rxjava2.imetate2

import java.util.concurrent.atomic.AtomicReference

/**
 * @author monk
 * @date 2019-03-01
 */
enum class DisposableHelper : Disposable {
    /***  */
    DISPOSABLE;

    override fun isDisposed(): Boolean {
        return false
    }

    override fun dispose() {}

    companion object {
        fun isDisposed(disposable: Disposable): Boolean {
            return disposable === DISPOSABLE
        }

        fun dispose(disposableAtomicReference: AtomicReference<Disposable?>): Boolean {
            var current = disposableAtomicReference.get()
            val d: Disposable = DISPOSABLE
            if (current !== d) {
                current = disposableAtomicReference.getAndSet(d)
                if (current !== d) {
                    current?.dispose()
                    return true
                }
            }
            return false
        }
    }
}